pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,263 @@
+package activator
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/apis/serving"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/prober"
+	"knative.dev/serving/pkg/queue"
+	"knative.dev/serving/pkg/reconciler"
+)
+
+type RevisionDestsUpdateFunc func(rev RevisionID, dests []string)
+
+const (
+	probePortName  string        = "http"
+	probeTimeout   time.Duration = 300 * time.Millisecond
+	probeFrequency time.Duration = 200 * time.Millisecond
+)
+
+type revisionWatcher struct {
+	rev          RevisionID
+	updateCb     RevisionDestsUpdateFunc
+	dests        []string
+	healthStates map[string]bool
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev RevisionID, updateCb RevisionDestsUpdateFunc,
+	destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:          rev,
+		updateCb:     updateCb,
+		healthStates: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		// TODO(greghaynes) log error here
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, net.JoinHostPort(addr.IP, strconv.Itoa(int(portVal))))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := []string{}
```

Comment:
but len(ret) != len(dests) here since we are filtering, its just at least as large. Are you saying to preallocate then chop down the slice? That might save a few appends but IMO were kind of splitting hairs at that point (this array really shouldnt get *huge* since we will switch to clusterIP in theory).

Reply:
I went ahead and updated this to overallocate then chop our slice down, IMO this is actually worse as its unlikely anyone can measure this perf difference but the logic is more complex but would rather not spin another review.

Reply:
@markusthoemmes tipped me off to how to do this differently using capacity separate from allocation.